### PR TITLE
gui: fix undefined behavior

### PR DESCRIPTION
**Pull request by Macdu**
It is undefined behavior to compare iterators of two different vectors.
It is undefined behavior to use the content of a variable outside of its living scope.

### CODE REVIEW

**Comment by bookmist** [comment]
There are other places where get_app_index is compared to some of `end()` iterators. For example: https://github.com/Vita3K/Vita3K/blob/fe7d6f01655b96d5ab1f8b998bbb0562b25a5193/vita3k/gui/src/main_menubar.cpp#L90

File: vita3k/gui/src/gui.cpp [hunk 1]
```diff
@@ -389,16 +388,14 @@ void init_app_background(GuiState &gui, EmuEnvState &emuenv, const std::string &
     else
         vfs::read_app_file(buffer, emuenv.pref_path.wstring(), app_path, "sce_sys/pic0.png");
 
-    const auto title = (APP_INDEX != gui.app_selector.sys_apps.end()) && (APP_INDEX != gui.app_selector.user_apps.end()) ? APP_INDEX->title : app_path;
```
Comment (Zangetsu38) [inline vita3k/gui/src/gui.cpp:392]: I no see any problèm here ?
Is just fine 🤔, same code is used also in other place

Comment (bookmist) [reply vita3k/gui/src/gui.cpp:392]: It crash on debug build at least.

Comment (Zangetsu38) [reply vita3k/gui/src/gui.cpp:392]: Is strange but ok 🤔

**Comment by Macdu** [comment]
> There are other places where get_app_index is compared to some of `end()` iterators. For example:
> 
> https://github.com/Vita3K/Vita3K/blob/fe7d6f01655b96d5ab1f8b998bbb0562b25a5193/vita3k/gui/src/main_menubar.cpp#L90

Yes you are right, I am going to make something a bit better instead then (return a pointer instead of an iterator)

**Comment by Macdu** [comment]
Ok, I modified the signature of get_app_index. That's not the cleanest thing (because some parts of the code assume get_app_index always return a value in user_apps) but it's better than what is being currently done and allows the debug builds no longer to crash.

File: vita3k/gui/src/content_manager.cpp [hunk 2]
```diff
@@ -103,7 +103,7 @@ static void get_save_data_list(GuiState &gui, EmuEnvState &emuenv) {
 
     for (const auto &save : fs::directory_iterator(SAVE_PATH)) {
         const auto title_id = save.path().stem().generic_string();
-        if (fs::is_directory(save.path()) && !fs::is_empty(save.path()) && get_app_index(gui, title_id) != gui.app_selector.user_apps.end()) {
+        if (fs::is_directory(save.path()) && !fs::is_empty(save.path()) && get_app_index(gui, title_id) != nullptr) {
```
Comment (Zangetsu38) [inline vita3k/gui/src/content_manager.cpp:106]:
```suggestion
        if (fs::is_directory(save.path()) && !fs::is_empty(save.path()) && get_app_index(gui, title_id)) {
```

Here like this should works too no ? You using this for title 🤔

File: vita3k/gui/src/main_menubar.cpp [hunk 3]
```diff
@@ -87,7 +87,7 @@ static void draw_emulation_menu(GuiState &gui, EmuEnvState &emuenv) {
                 for (auto i = 0; i < std::min(8, int32_t(gui.time_apps[emuenv.io.user_id].size())); i++) {
                     const auto time_app = gui.time_apps[emuenv.io.user_id][i];
                     const auto app_index = get_app_index(gui, time_app.app);
-                    if ((app_index != gui.app_selector.user_apps.end()) && (app_index != gui.app_selector.sys_apps.end()))
+                    if (app_index != nullptr)
```
Comment (Zangetsu38) [inline vita3k/gui/src/main_menubar.cpp:90]:
```suggestion
                    if (app_index)
```
Should he same here

File: vita3k/gui/src/content_manager.cpp (hunk 2 above)
Comment (Macdu) [reply vita3k/gui/src/content_manager.cpp:106]: Yes, this also works, but I feel like the != nullptr is a bit more expressive (you get to know that get_app_index returns a pointer and not a boolean). Although that's only personal preference.

Comment (Zangetsu38) [reply vita3k/gui/src/content_manager.cpp:106]: Vita3k more no using it, dev have some time refactor for remove all of this with compare on nullptr 
Is more clean

**Review by Zangetsu38 — APPROVED** [review]

**Review by bookmist — APPROVED** [review]

**Pull request merged**